lm.totalTableSize underflow

Apply this diff:

```diff
 func (lm *LevelMetadata) remove(f *TableMetadata) {
+       if lm.totalTableSize < f.Size {
+               fmt.Printf("f: %v\nin tree:\n", f)
+               for t := range lm.tree.All() {
+                       fmt.Printf("  - %v\n", t)
+               }
+               panic("foo")
+       }
```

```
go test -tags invariants .  |& tee /tmp/testlog
f: 000005:[a#10,SET-c#12,SET]
in tree:
panic: foo

goroutine 1730 [running]:
github.com/cockroachdb/pebble/internal/manifest.(*LevelMetadata).remove(0x140004508b0, 0x2?)
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/manifest/level_metadata.go:106 +0x19c
github.com/cockroachdb/pebble/internal/manifest.(*BulkVersionEdit).Apply(0x1400101b490, 0x14000450388, 0x3e80)
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/manifest/version_edit.go:1362 +0xc7c
github.com/cockroachdb/pebble.(*versionSet).UpdateVersionLocked.func1(0x14000264008, 0x14000246dd0, 0x0, 0x8, 0xb, {0x0, 0x0, 0x0}, {0x14000246dd0, 0xb, ...}, ...)
        /Users/radu/go/src/github.com/cockroachdb/pebble/version_set.go:501 +0x188
github.com/cockroachdb/pebble.(*versionSet).UpdateVersionLocked(0x14000264008, 0x1400101bd08)
        /Users/radu/go/src/github.com/cockroachdb/pebble/version_set.go:538 +0x4f8
github.com/cockroachdb/pebble.(*DB).compact1(0x140003ce008, 0xb, 0x14000e4a688)
        /Users/radu/go/src/github.com/cockroachdb/pebble/compaction.go:2730 +0x254
github.com/cockroachdb/pebble.(*tableCompaction).Execute(0x14000e4a688, 0xb, 0x140003ce008)
        /Users/radu/go/src/github.com/cockroachdb/pebble/compaction.go:385 +0x44
github.com/cockroachdb/pebble.(*DB).compact.func1.1(0x140003ce008, {0x10142caa8, 0x14000e4a688}, 0x140006c7730)
        /Users/radu/go/src/github.com/cockroachdb/pebble/compaction.go:2592 +0xe4
github.com/cockroachdb/pebble.(*DB).compact.func1({0x101423a78?, 0x14000243860?})
        /Users/radu/go/src/github.com/cockroachdb/pebble/compaction.go:2621 +0x38
runtime/pprof.Do({0x1014234b8?, 0x101a5a660?}, {{0x14000522b40?, 0x0?, 0x0?}}, 0x14000a41f70)
        /Users/radu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.5.darwin-arm64/src/runtime/pprof/runtime.go:51 +0x78
github.com/cockroachdb/pebble.(*DB).compact(0x140003ce008, {0x10142caa8, 0x14000e4a688}, 0x140006c7730)
        /Users/radu/go/src/github.com/cockroachdb/pebble/compaction.go:2586 +0xd0
github.com/cockroachdb/pebble.(*DB).runPickedCompaction.func1()
        /Users/radu/go/src/github.com/cockroachdb/pebble/compaction.go:2112 +0x2c
created by github.com/cockroachdb/pebble.(*DB).runPickedCompaction in goroutine 1699
        /Users/radu/go/src/github.com/cockroachdb/pebble/compaction.go:2111 +0x110
FAIL    github.com/cockroachdb/pebble   1.004s
```

--- a/internal/manifest/btree.go
+++ b/internal/manifest/btree.go
@@ -736,13 +736,15 @@ func (t *btree[M]) Clone() btree[M] {
 // Delete removes the provided table from the tree, unrefing the table's files
 // if it's found. If any files are unreferenced to zero, they're added to the
 // provided obsoleteFiles.
-func (t *btree[M]) Delete(item M, of ObsoleteFilesSet) {
+func (t *btree[M]) Delete(item M, of ObsoleteFilesSet) (found bool) {
 	if t.root == nil || t.root.count == 0 {
-		return
+		return false
 	}
-	if out, found := mut(&t.root).Remove(t.bcmp, item); found {
+	out, found := mut(&t.root).Remove(t.bcmp, item)
+	if found {
 		out.Unref(of)
 	}
+	// Note that even if the item was not found, we could have rebalanced a node.
 	if invariants.Enabled {
 		t.root.verifyInvariants()
 	}
@@ -755,6 +757,7 @@ func (t *btree[M]) Delete(item M, of ObsoleteFilesSet) {
 		}
 		old.decRef(false /* contentsToo */, assertNoObsoleteFiles{})
 	}
+	return found
 }
 
 // Insert adds the given item to the tree. If a item in the tree already

--- a/internal/manifest/level_metadata.go
+++ b/internal/manifest/level_metadata.go
@@ -12,6 +12,7 @@ import (
 
 	"github.com/cockroachdb/pebble/internal/base"
 	"github.com/cockroachdb/pebble/internal/invariants"
+	"github.com/cockroachdb/pebble/metrics"
 )
 
 // LevelMetadata contains metadata for all of the files within
@@ -20,23 +21,19 @@ type LevelMetadata struct {
 	level          int
 	totalTableSize uint64
 	totalRefSize   uint64
-	// NumVirtual is the number of virtual sstables in the level.
-	NumVirtual uint64
-	// VirtualTableSize is the size of the virtual sstables in the level.
-	VirtualTableSize uint64
-	tree             btree[*TableMetadata]
+	virtualTables  metrics.CountAndSize
+	tree           btree[*TableMetadata]
 }
 
 // clone makes a copy of the level metadata, implicitly increasing the ref
 // count of every file contained within lm.
 func (lm *LevelMetadata) clone() LevelMetadata {
 	return LevelMetadata{
-		level:            lm.level,
-		totalTableSize:   lm.totalTableSize,
-		totalRefSize:     lm.totalRefSize,
-		NumVirtual:       lm.NumVirtual,
-		VirtualTableSize: lm.VirtualTableSize,
-		tree:             lm.tree.Clone(),
+		level:          lm.level,
+		totalTableSize: lm.totalTableSize,
+		totalRefSize:   lm.totalRefSize,
+		virtualTables:  lm.virtualTables,
+		tree:           lm.tree.Clone(),
 	}
 }
 
@@ -57,8 +54,7 @@ func MakeLevelMetadata(cmp Compare, level int, files []*TableMetadata) LevelMeta
 		lm.totalTableSize += f.Size
 		lm.totalRefSize += f.EstimatedReferenceSize()
 		if f.Virtual {
-			lm.NumVirtual++
-			lm.VirtualTableSize += f.Size
+			lm.virtualTables.Inc(f.Size)
 		}
 	}
 	return lm
@@ -91,20 +87,21 @@ func (lm *LevelMetadata) insert(f *TableMetadata) error {
 	lm.totalTableSize += f.Size
 	lm.totalRefSize += f.EstimatedReferenceSize()
 	if f.Virtual {
-		lm.NumVirtual++
-		lm.VirtualTableSize += f.Size
+		lm.virtualTables.Inc(f.Size)
 	}
 	return nil
 }
 
-func (lm *LevelMetadata) remove(f *TableMetadata) {
-	lm.totalTableSize -= f.Size
-	lm.totalRefSize -= f.EstimatedReferenceSize()
-	if f.Virtual {
-		lm.NumVirtual--
-		lm.VirtualTableSize -= f.Size
+func (lm *LevelMetadata) remove(f *TableMetadata) (found bool) {
+	if !lm.tree.Delete(f, assertNoObsoleteFiles{}) {
+		return false
 	}
-	lm.tree.Delete(f, assertNoObsoleteFiles{})
+	lm.totalTableSize = invariants.SafeSub(lm.totalTableSize, f.Size)
+	lm.totalRefSize = invariants.SafeSub(lm.totalRefSize, f.EstimatedReferenceSize())
+	if f.Virtual {
+		lm.virtualTables.Dec(f.Size)
+	}
+	return true
 }
 
 // Empty indicates whether there are any files in the level.
@@ -135,6 +132,12 @@ func (lm *LevelMetadata) TableSize() uint64 {
 // AggregateSize().
 func (lm *LevelMetadata) EstimatedReferenceSize() uint64 {
 	return lm.totalRefSize
+}
+
+// VirtualTables returns the count and total size of the virtual tables in the
+// level.
+func (lm *LevelMetadata) VirtualTables() metrics.CountAndSize {
+	return lm.virtualTables
 }
 
 // Iter constructs a LevelIterator over the entire level.

--- a/internal/manifest/version.go
+++ b/internal/manifest/version.go
@@ -102,6 +102,10 @@ func NewVersionForTesting(
 		}
 		for _, f := range files[l] {
 			v.Levels[l].totalTableSize += f.Size
+			v.Levels[l].totalRefSize += f.EstimatedReferenceSize()
+			if f.Virtual {
+				v.Levels[l].virtualTables.Inc(f.Size)
+			}
 		}
 	}
 	l0Organizer.ResetForTesting(v)

--- a/internal/manifest/version_edit.go
+++ b/internal/manifest/version_edit.go
@@ -1358,7 +1358,11 @@ func (b *BulkVersionEdit) Apply(curr *Version, readCompactionRate int64) (*Versi
 			// counts.  However, because we cloned the previous level's B-Tree,
 			// this should never result in a file's reference count dropping to
 			// zero. The remove call will panic if this happens.
-			v.Levels[level].remove(f)
+			if !v.Levels[level].remove(f) {
+				if invariants.Enabled {
+					panic(errors.AssertionFailedf("deleted table %f not in level", f))
+				}
+			}
 			v.RangeKeyLevels[level].remove(f)
 			v.MarkedForCompaction.Delete(f, level)
 		}

--- a/version_set.go
+++ b/version_set.go
@@ -1082,8 +1082,8 @@ func setBasicLevelMetrics(m *Metrics, newVersion *manifest.Version) {
 		l := &m.Levels[i]
 		l.TablesCount = int64(newVersion.Levels[i].Len())
 		l.TablesSize = int64(newVersion.Levels[i].TableSize())
-		l.VirtualTablesCount = newVersion.Levels[i].NumVirtual
-		l.VirtualTablesSize = newVersion.Levels[i].VirtualTableSize
+		l.VirtualTablesCount = newVersion.Levels[i].VirtualTables().Count
+		l.VirtualTablesSize = newVersion.Levels[i].VirtualTables().Bytes
 		l.EstimatedReferencesSize = newVersion.Levels[i].EstimatedReferenceSize()
 		l.Sublevels = 0
 		if l.TablesCount > 0 {

